python/pyarrow/tests/test_dataset.py

```diff
@@ -4192,27 +4192,27 @@ def test_write_table_multiple_fragments(tempdir):
     # Table with multiple batches written as single Fragment by default
     base_dir = tempdir / 'single'
     ds.write_dataset(table, base_dir, format="feather")
-    assert set(base_dir.rglob("*")) == set([base_dir / "part-0.feather"])
+    assert set(base_dir.rglob("*")) == set([base_dir / "part-0.arrow"])
     assert ds.dataset(base_dir, format="ipc").to_table().equals(table)
 
     # Same for single-element list of Table
     base_dir = tempdir / 'single-list'
     ds.write_dataset([table], base_dir, format="feather")
-    assert set(base_dir.rglob("*")) == set([base_dir / "part-0.feather"])
+    assert set(base_dir.rglob("*")) == set([base_dir / "part-0.arrow"])
     assert ds.dataset(base_dir, format="ipc").to_table().equals(table)
 
     # Provide list of batches to write multiple fragments
     base_dir = tempdir / 'multiple'
     ds.write_dataset(table.to_batches(), base_dir, format="feather")
     assert set(base_dir.rglob("*")) == set(
-        [base_dir / "part-0.feather"])
+        [base_dir / "part-0.arrow"])
```

Comment:
is this the default even if format is specified? (eg feather would still be appropriate here)

Reply:
Since this project recommends `.arrow` as the extension for files created with this format, I feel it is necessary to use `.arrow` here for consistency.

https://github.com/apache/arrow/blob/1b9c57e20802fb061c90837c39e99d8fa69cc212/docs/source/format/Columnar.rst#L1047-L1052

@kou Do you have any additional explanation?

Reply:
Ah, it's a good point.
If users specify `"feather"` explicitly, it may be better that we use `".feather"` for extension name. Does this work?

```diff
diff --git a/python/pyarrow/_dataset.pyx b/python/pyarrow/_dataset.pyx
index 93d3555b94..f52f8a6519 100644
--- a/python/pyarrow/_dataset.pyx
+++ b/python/pyarrow/_dataset.pyx
@@ -1154,6 +1154,13 @@ cdef class IpcFileFormat(FileFormat):
         return IpcFileFormat, tuple()
 
 
+cdef class FeatherFileFormat(IpcFileFormat):
+
+    @property
+    def default_extname(self):
+        return "feather"
+
+
 cdef class CsvFileFormat(FileFormat):
     """
     FileFormat for CSV files.
diff --git a/python/pyarrow/dataset.py b/python/pyarrow/dataset.py
index 326b37ec6e..5816bd54ae 100644
--- a/python/pyarrow/dataset.py
+++ b/python/pyarrow/dataset.py
@@ -284,8 +284,10 @@ def _ensure_format(obj):
         if not _parquet_available:
             raise ValueError(_parquet_msg)
         return ParquetFileFormat()
-    elif obj in {"ipc", "arrow", "feather"}:
+    elif obj in {"ipc", "arrow"}:
         return IpcFileFormat()
+    elif obj in {"feather"}:
+        return FeatherFileFormat()
     elif obj == "csv":
         return CsvFileFormat()
     elif obj == "orc":
```

Reply:
Hmmm, I wonder if it is worth complicating the source code just to keep the feather extension. (Of course, I'm not the maintainer, so if that's ok with the maintainer, fine.)

Considering that the extension can be changed to any extension by setting the `basename_template` option (as @westonpace says), I think there is no need to treat only the `"feather"` case specially here.

I want to emphasize that it is hard to understand the relationship between IPC files and Feather files anyway.
For example, in Julia, if we want to read an IPC file, I need Arrow.jl, but if I want to read a Feather V1 file, we need the Feather.jl library.

Reply:
I don't think that the implementation I suggested is so complex (if it works).

I think that specifying `format="feather"` explicitly means that "the user wants to call the format Feather V2 not Apache Arrow IPC file format" because the user can specify `format="ipc"` or `format="arrow"` instead of `format="feather"`. I don't know why we need to force the user to specify `basename_template` too for this case.

> I want to emphasize that it is hard to understand the relationship between IPC files and Feather files anyway.

We may need to deprecate Feather V2. Could you start a discussion on the `dev@arrow.apache.org` mailing list?

> For example, in Julia, if we want to read an IPC file, I need Arrow.jl, but if I want to read a Feather V1 file, we need the Feather.jl library.

We can resolve the problem by implementing auto-detection feature like we did in Apache Arrow C++: https://github.com/apache/arrow/blob/1b9c57e20802fb061c90837c39e99d8fa69cc212/cpp/src/arrow/ipc/feather.cc#L785-L797

I'm not sure where the auto-detection feature should be implemented (Arrow.jl, Feather.jl or new library?) but how about create an issue to https://github.com/apache/arrow-julia ?

Reply:
> I think that specifying `format="feather"` explicitly means that "the user wants to call the format Feather V2 not Apache Arrow IPC file format" because the user can specify `format="ipc"` or `format="arrow"` instead of `format="feather"`.

I don't know why there are multiple aliases set up here, but as long as this project have treated Feather V2 as an alias for IPC files in the past (or still do), I don't know if users are choosing `"feather"` for a clear reason.
For example, we use the `pyarrow.feather` module to handle IPC files with pyarrow. (not `pyarrow.ipc` or `pyarrow.arrow`.)

> We may need to deprecate Feather V2. Could you start a discussion on the `dev@arrow.apache.org` mailing list?

To be clear, I just think it makes more sense to use the extension `.arrow` for Feather V2 files, not that I think the name "Feather V2" should be discontinued.

I had assumed that such a policy was adopted once the official extension for IPC files was decided to be `.arrow`, but apparently not?
(BTW, I was surprised at the time because I thought `.feather` was the official extension until I saw [the news](https://www.clear-code.com/blog/2022/5/13/latest-apache-arrow-information.html).)

For now, it may make sense here to leave the `.feather` extension for the `"feather"` case, and warn in the future?

> I'm not sure where the auto-detection feature should be implemented (Arrow.jl, Feather.jl or new library?) but how about create an issue to [apache/arrow-julia](https://github.com/apache/arrow-julia) ?

Thanks for your suggestion, but Julia's case is an example, not that I want that feature.
I just wanted to say that there may be no benefit in continuing to use the `.feather` extension, which may not be readable by some libraries.

Reply:
> For example, we use the `pyarrow.feather` module to handle IPC files with pyarrow. (not `pyarrow.ipc` or `pyarrow.arrow`.)

No. Normally, users use `pyarrow.ipc.open_file()`/`pyarrow.ipc.new_file()` for it. See also:

* https://arrow.apache.org/docs/python/ipc.html#writing-and-reading-random-access-files
* https://arrow.apache.org/cookbook/py/io.html#saving-arrow-arrays-to-disk

> For now, it may make sense here to leave the `.feather` extension for the `"feather"` case, and warn in the future?

@westonpace What do you think about this?

Reply:
Also @jorisvandenbossche for opinions.

Reply:
Personally, I am +1 on the suggestion of @kou for having a small FeatherFileFormat class so that we can let the default extension name depend on whether the user specified `format="ipc"` vs `format="feather"` (for the reasons @kou gave).  
I am only wondering if we can avoid making that public, so it's only used under the hood if the user passed a string format name to write_dataset.

(I also do think it might be worth a broader discussion about to which extent and how we want to keep using the Feather name)

Reply:
> No. Normally, users use `pyarrow.ipc.open_file()`/`pyarrow.ipc.new_file()` for it. See also:

Oh...... I really did not know that.

This is the part I was actually looking at because I believed these references were classified as "Tabular File Formats".

![image](https://user-images.githubusercontent.com/50911393/187181281-da96cca9-64a5-462f-9b41-4c46db3edd1e.png)